src/nvapi/nvapi_d3d11_device.cpp

```diff
@@ -1,66 +1,131 @@
 
 namespace dxvk {
-    bool NvapiD3d11Device::SetDepthBoundsTest(IUnknown* deviceOrContext, const bool enable, const float minDepth, const float maxDepth) {
-        auto sdbtDeviceContext = GetSdbtDeviceContext(deviceOrContext);
-        if (sdbtDeviceContext == nullptr)
+    std::unordered_map<IUnknown*, std::shared_ptr<NvapiD3d11Device>> NvapiD3d11Device::m_nvapiDeviceMap;
+    std::mutex NvapiD3d11Device::m_mutex;
+
+    void NvapiD3d11Device::Reset() {
+        std::scoped_lock lock{m_mutex};
+        m_nvapiDeviceMap.clear();
+    }
+
+    std::pair<Com<ID3D11VkExtDevice>, Com<ID3D11VkExtContext>> GetDxvkDevice(IUnknown* device) {
+        Com<ID3D11Device> d3d11Device;
+        Com<ID3D11DeviceContext> d3d11DeviceContext;
+
+        if (SUCCEEDED(device->QueryInterface(IID_PPV_ARGS(&d3d11Device))))
+            d3d11Device->GetImmediateContext(&d3d11DeviceContext);
+        else if (SUCCEEDED(device->QueryInterface(IID_PPV_ARGS(&d3d11DeviceContext))))
+            d3d11DeviceContext->GetDevice(&d3d11Device);
+
+        if (d3d11Device == nullptr || d3d11DeviceContext == nullptr)
+            return {};
+
+        Com<ID3D11VkExtDevice> dxvkDevice;
+        Com<ID3D11VkExtContext> dxvkContext;
+        if (SUCCEEDED(d3d11Device->QueryInterface(IID_PPV_ARGS(&dxvkDevice))) && SUCCEEDED(d3d11DeviceContext->QueryInterface(IID_PPV_ARGS(&dxvkContext))))
+            return std::make_pair(dxvkDevice, dxvkContext);
+
+        return {};
+    }
+
+    NvapiD3d11Device* NvapiD3d11Device::GetOrCreate(IUnknown* device) {
+        std::scoped_lock lock{m_mutex};
+
+        if (auto nvapiD3D11Device = Get(device))
+            return nvapiD3D11Device;
+
+        auto [dxvkDevice, dxvkContext] = GetDxvkDevice(device);
+        if (dxvkDevice == nullptr || dxvkContext == nullptr)
+            return nullptr;
+
+        // Look for a cache entry where NvapiD3d11Device's m_dxvkDevice and m_dxvkContext matches the ones we found
+        auto itF = std::find_if(m_nvapiDeviceMap.begin(), m_nvapiDeviceMap.end(),
+            [&dxvkDevice, &dxvkContext](auto& item) { return item.second->m_dxvkDevice == dxvkDevice.ptr() && item.second->m_dxvkContext == dxvkContext.ptr(); });
+
+        auto [itI, inserted] = itF == m_nvapiDeviceMap.end()
+            ? m_nvapiDeviceMap.emplace(device, std::make_shared<NvapiD3d11Device>(dxvkDevice.ptr(), dxvkContext.ptr()))
+            : m_nvapiDeviceMap.emplace(device, itF->second);
+
+        if (!inserted)
+            return nullptr;
+
+        return itI->second.get();
+    }
+
+    NvapiD3d11Device* NvapiD3d11Device::Get(IUnknown* device) {
+        auto it = m_nvapiDeviceMap.find(device);
+        return it == m_nvapiDeviceMap.end() ? nullptr : it->second.get();
+    }
+
+    NvapiD3d11Device::NvapiD3d11Device(ID3D11VkExtDevice* dxvkDevice, ID3D11VkExtContext* dxvkContext)
+        : m_dxvkDevice(dynamic_cast<ID3D11VkExtDevice1*>(dxvkDevice)), m_dxvkContext(dynamic_cast<ID3D11VkExtContext1*>(dxvkContext)) {
```

Comment:
This must be `static_cast`, `dynamic_cast` returns `nullptr` if this isn't a *1 device/context.